core/drivers/pl022_spi.c

```diff
@@ -0,0 +1,505 @@
+
+static void pl022_calc_clk_divisors(struct pl022_data *pd,
+	uint8_t *cpsdvr, uint8_t *scr)
+{
+	unsigned int freq1 = 0, freq2 = 0;
+	uint8_t tmp_cpsdvr1, tmp_scr1, tmp_cpsdvr2 = 0, tmp_scr2 = 0;
+
+	for (tmp_scr1 = SSP_SCR_MIN; tmp_scr1 < SSP_SCR_MAX; tmp_scr1++) {
+		for (tmp_cpsdvr1 = SSP_CPSDVR_MIN; tmp_cpsdvr1 < SSP_CPSDVR_MAX;
+			tmp_cpsdvr1++) {
+			freq1 = pl022_calc_freq(pd, tmp_cpsdvr1, tmp_scr1);
+			if (freq1 == pd->speed_hz)
+				goto done;
+			else if (freq1 < pd->speed_hz)
+				goto stage2;
+		}
+	}
+
+stage2:
+	for (tmp_cpsdvr2 = SSP_CPSDVR_MIN; tmp_cpsdvr2 < SSP_CPSDVR_MAX;
+		tmp_cpsdvr2++) {
+		for (tmp_scr2 = SSP_SCR_MIN; tmp_scr2 < SSP_SCR_MAX;
+			tmp_scr2++) {
+			freq2 = pl022_calc_freq(pd, tmp_cpsdvr2, tmp_scr2);
+			if (freq2 <= pd->speed_hz)
+				goto done;
+		}
+	}
+
+done:
+	if (freq1 >= freq2) {
+		*cpsdvr = tmp_cpsdvr1;
+		*scr = tmp_scr1;
+		DMSG("speed: requested: %u, closest1: %u\n",
+			pd->speed_hz, freq1);
+	} else {
+		*cpsdvr = tmp_cpsdvr2;
+		*scr = tmp_scr2;
+		DMSG("speed: requested: %u, closest2: %u\n",
+			pd->speed_hz, freq2);
+	}
+	DMSG("cpsdvr: %u (0x%x), scr: %u (0x%x)\n",
+		*cpsdvr, *cpsdvr, *scr, *scr);
+}
+
+static void pl022_flush_fifo(struct pl022_data *pd)
+{
+	uint32_t rdat;
+
+	do {
+		while (read32(pd->base + SSPSR) & SSPSR_RNE) {
+			rdat = read32(pd->base + SSPDR);
+			DMSG("rdat: 0x%x\n", rdat);
+		}
+	} while (read32(pd->base + SSPSR) & SSPSR_BSY);
+}
+
+static const struct spi_ops pl022_ops = {
+	.txrx8 = pl022_txrx8,
+	.txrx16 = pl022_txrx16,
+	.tx8 = pl022_tx8,
+	.tx16 = pl022_tx16,
+	.rx8 = pl022_rx8,
+	.rx16 = pl022_rx16,
+};
+
+void pl022_configure(struct pl022_data *pd)
+{
+	uint16_t mode, data_size;
+	uint8_t cpsdvr, scr, lbm;
+
+	pd->chip.ops = &pl022_ops;
+	pl022_sanity_check(pd);
+	pl022_calc_clk_divisors(pd, &cpsdvr, &scr);
+
+	/* configure ssp based on platform settings */
+	switch (pd->mode) {
+	case SPI_MODE0:
+		DMSG("SPI_MODE0\n");
+		mode = PL022_SPI_MODE0;
+		break;
+	case SPI_MODE1:
+		DMSG("SPI_MODE1\n");
+		mode = PL022_SPI_MODE1;
+		break;
+	case SPI_MODE2:
+		DMSG("SPI_MODE2\n");
+		mode = PL022_SPI_MODE2;
+		break;
+	case SPI_MODE3:
+		DMSG("SPI_MODE3\n");
+		mode = PL022_SPI_MODE3;
+		break;
+	default:
+		EMSG("Invalid spi mode: %u\n", pd->mode);
+		panic();
+	}
+
+	switch (pd->data_size_bits) {
+	case 8:
+		DMSG("Data size: 8\n");
+		data_size = PL022_DATA_SIZE8;
+		break;
+	case 16:
+		DMSG("Data size: 16\n");
+		data_size = PL022_DATA_SIZE16;
+		break;
+	default:
+		EMSG("Unsupported data size: %u bits\n", pd->data_size_bits);
+		panic();
+	}
+
+	if (pd->loopback) {
+		DMSG("Starting in loopback mode!\n");
+		lbm = SSPCR1_LBM_YES;
+	} else {
+		DMSG("Starting in regular (non-loopback) mode!\n");
+		lbm = SSPCR1_LBM_NO;
+	}
+
+	DMSG("set serial clock rate (scr), spi mode (phase and clock)\n");
+	DMSG("set frame format (spi) and data size (8- or 16-bit)\n");
+	io_mask16(pd->base + SSPCR0, SHIFT_U32(scr, 8) | mode | SSPCR0_FRF_SPI
+		| data_size, MASK_16);
```

Comment:
minor coding style issue

the `|` (bitwise or) here should appear at the end of the upper line, rather than at start of the next line. This coding style directive stand for all operations: `+` `-` `&` `&&` and other operators.

for example:
```
if (condition_a_is_true &&
    condition_b_is_true)
	do_something();

value = (data & mask) | 
		new_value;
```